ruma-common/src/push.rs

```diff
@@ -88,126 +87,101 @@ impl Ruleset {
-        #[derive(Clone, Debug, Serialize, Deserialize)]
-        #[doc = "Wrapper type to disambiguate the kind of the wrapped rule"]
-        pub struct $name(pub $inner);
-
-        impl RulesetMember for $name {
-            fn add_to(self, ruleset: &mut Ruleset) -> bool {
-                ruleset.$field.insert(self)
-            }
-        }
+/// The kinds of push rules that are available.
+#[derive(Clone, Debug)]
+pub enum AnyPushRule {
+    /// Rules that override all other kinds.
+    Override(ConditionalPushRule),
 
-        impl Extend<$name> for Ruleset {
-            fn extend<T: IntoIterator<Item = $name>>(&mut self, iter: T) {
-                for rule in iter {
-                    rule.add_to(self);
-                }
-            }
-        }
+    /// Content-specific rules.
+    Content(PatternedPushRule),
 
-        // The following trait are needed to be able to make
-        // an IndexSet of the new type
+    /// Room-specific rules.
+    Room(SimplePushRule),
 
-        impl Hash for $name {
-            fn hash<H: Hasher>(&self, state: &mut H) {
-                self.0.rule_id.hash(state);
-            }
-        }
+    /// Sender-specific rules.
+    Sender(SimplePushRule),
 
-        impl PartialEq for $name {
-            fn eq(&self, other: &Self) -> bool {
-                self.0.rule_id == other.0.rule_id
-            }
-        }
+    /// Lowest priority rules.
+    Underride(ConditionalPushRule),
+}
 
-        impl Eq for $name {}
+impl AnyPushRule {
+    /// Adds a value in the correct field of a `Ruleset`.
+    #[doc(hidden)]
+    fn add_to(self, ruleset: &mut Ruleset) -> bool {
+        match self {
+            Self::Override(rule) => ruleset.override_.insert(rule),
+            Self::Underride(rule) => ruleset.underride.insert(rule),
+            Self::Content(rule) => ruleset.content.insert(rule),
+            Self::Room(rule) => ruleset.room.insert(rule),
+            Self::Sender(rule) => ruleset.sender.insert(rule),
+        }
+    }
 
-        impl Equivalent<$name> for str {
-            fn equivalent(&self, key: &$name) -> bool {
-                self == key.0.rule_id
-            }
+    /// The `rule_id` of the push rule
+    pub fn rule_id(&self) -> String {
+        match self {
+            Self::Override(rule) => rule.rule_id.clone(),
+            Self::Underride(rule) => rule.rule_id.clone(),
+            Self::Content(rule) => rule.rule_id.clone(),
+            Self::Room(rule) => rule.rule_id.clone(),
+            Self::Sender(rule) => rule.rule_id.clone(),
         }
-    };
+    }
 }
 
-rulekind!(OverridePushRule, ConditionalPushRule, override_);
-rulekind!(UnderridePushRule, ConditionalPushRule, underride);
-rulekind!(RoomPushRule, SimplePushRule, room);
-rulekind!(SenderPushRule, SimplePushRule, sender);
-rulekind!(ContentPushRule, PatternedPushRule, content);
+impl Extend<AnyPushRule> for Ruleset {
+    fn extend<T: IntoIterator<Item = AnyPushRule>>(&mut self, iter: T) {
```

Comment:
Can you move the `IntoIterator` bound into a where clause like in the trait declaration? (just for readability)